### PR TITLE
Use separable filter2d for box filter

### LINKED ISSUE #2246: Improve speed for box blur (mean filter)
## 🚀 Feature

Originally from this: https://github.com/kornia/kornia/issues/720#issuecomment-881859300

In theory box blur (mean filter) can be implemented in O(n) time.

Two solutions

- Use 2 separable 1D kernels (like `gaussian_blur2d()`) -> O(nk) time, should be simple. Only launch 2 pytorch calls.
- Use intergral image (like [here](https://github.com/wuhuikai/DeepGuidedFilter/blob/master/GuidedFilteringLayer/GuidedFilter_PyTorch/guided_filter_pytorch/box_filter.py)) -> O(n) time. It seems like there will be quite a lot of overhead for small kernels

## Additional context

______________________________________________________________________

#### Consider also to contribute to Kornia universe projects :)

<sub>

- [**Tutorials**](https://github.com/kornia/tutorials): our repository containing the tutorials.

</sub>

### PR DESCRIPTION
**Pull request by gau-nernst**
#### Changes

Fixes #2246 

Following separable filter implementation of Gaussian blur, I add separable filter2d for Box blur.

When I benchmark the speed, it is clear that both kernel size and image size must be large to observe speedup. (device is RTX 3090, PyTorch 2.0, CUDA 11.8)

box_blur_rtx3090_128x128

box_blur_rtx3090_1024x1024

I also ran the speed benchmark on Mac M1 CPU (PyTorch 2.0). The results are... interesting. At large batch sizes, separable kernels don't help much. `torch.compile()` with 3x3 filter2d is super fast.

box_blur

Benchmark script

```python
import itertools
import time

import matplotlib.pyplot as plt
import pandas as pd
import torch

from kornia.filters import box_blur


N = 20
device = "cuda"
img_dims = (3, 128, 128)

batch_sizes = [1, 10, 100]
kernel_sizes = [3, 5, 7, 9, 11]
separables = [False, True]
do_compiles = [False, True]

samples = []

for batch_size, kernel_size, separable, do_compile in itertools.product(
    batch_sizes, kernel_sizes, separables, do_compiles
):
    img = torch.randn(batch_size, *img_dims, device=device)
    f = torch.compile(box_blur) if do_compile else box_blur
    for _ in range(5):
        out = f(img, kernel_size, separable=separable)

    torch.cuda.synchronize()
    time0 = time.perf_counter()
    for _ in range(N):
        torch.cuda.synchronize()
        out = f(img, kernel_size, separable=separable)
        torch.cuda.synchronize()
    throughput = N / (time.perf_counter() - time0)

    samples.append((batch_size, kernel_size, separable, do_compile, throughput))
    print(*samples[-1], sep="\t")

df = pd.DataFrame(samples, columns=["batch size", "kernel size", "separable", "torch.compile", "throughput"])
df = df.pivot(index=["batch size", "kernel size"], columns=["separable", "torch.compile"], values="throughput")
df.values[:, :] /= df.values[:, [0]]
print(df)

df.plot.bar(rot=0, figsize=(12, 8))
plt.tight_layout()
plt.savefig("box_blur.png")
```

**Questions**:

- Perhaps I should leave the default value for `separable` as `False`?
- I notice that `normalized` does not do anything since the kernel is already normalized, thus I don't use the parameter `normalized` at all. Shall we deprecate the use of `normalized` parameter here? i.e. raise DeprecationWarning.

#### Type of change

- [ ] 📚  Documentation Update
- [ ] 🧪 Tests Cases
- [ ] 🐞 Bug fix (non-breaking change which fixes an issue)
- [x] 🔬 New feature (non-breaking change which adds functionality)
- [ ] 🚨 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📝 This change requires a documentation update

#### Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] Did you update CHANGELOG in case of a major change?

### CODE REVIEW

**Comment by edgarriba** [comment]
interesting results ! 

> I notice that normalized does not do anything since the kernel is already normalized, thus I don't use the parameter normalized at all. Shall we deprecate the use of normalized parameter here? i.e. raise DeprecationWarning.

i believe this was @ducha-aiki

**Comment by ducha-aiki** [comment]
Great job!
Yes, let's set `separable=False` by default. And let's remove the normalization, as the kernel is generated by ourselves and not provided as an input.

File: kornia/filters/blur.py [hunk 1]
```diff
@@ -87,20 +92,33 @@ class BoxBlur(Module):
...
-            f"border_type={self.border_type})"
+            f"border_type={self.border_type}, "
+            f"separable={self.separable})"
         )
 
     def forward(self, input: Tensor) -> Tensor:
-        return box_blur(input, self.kernel_size, self.border_type, self.normalized)
+        if self.separable:
+            return filter2d_separable(input, self.kernel_x, self.kernel_y, self.border_type)
+        else:
```
Comment (edgarriba) [inline kornia/filters/blur.py:123]:
```suggestion
```

File: kornia/filters/blur.py [hunk 2]
```diff
@@ -87,20 +92,33 @@ class BoxBlur(Module):
...
+            return filter2d(input, self.kernel, self.border_type)
```
Comment (edgarriba) [inline kornia/filters/blur.py:124]:
```suggestion
        return filter2d(input, self.kernel, self.border_type)
```

File: kornia/filters/kernels.py [hunk 3]
```diff
@@ -260,6 +260,12 @@ def laplacian_1d(window_size: int, *, device: Device | None = None, dtype: Dtype
     return filter_1d
 
 
+def get_box_kernel1d(kernel_size: int, *, device: Device | None = None, dtype: Dtype | None = None) -> Tensor:
```
Comment (edgarriba) [inline kornia/filters/kernels.py:263]: add docstrigns for args

**Review by ducha-aiki — APPROVED** [review]

**Review by johnnv1 — APPROVED** [review]
